src/mongoose_cets_discovery_rdbms.erl

```diff
@@ -5,52 +5,99 @@
 
 -include_lib("kernel/include/logger.hrl").
 
--type opts() :: #{cluster_name => binary(), node_name_to_insert => binary()}.
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+-type opts() :: #{cluster_name => binary(), node_name_to_insert => binary(), last_query_info => term()}.
 -type state() :: opts().
 
 -spec init(opts()) -> state().
 init(Opts = #{cluster_name := _, node_name_to_insert := _}) ->
-    Opts.
+    Opts#{last_query_info => #{}}.
 
 -spec get_nodes(state()) -> {cets_discovery:get_nodes_result(), state()}.
 get_nodes(State = #{cluster_name := ClusterName, node_name_to_insert := Node}) ->
-    prepare(),
-    insert(ClusterName, Node),
-    try mongoose_rdbms:execute_successfully(global, cets_disco_select, [ClusterName]) of
-        {selected, Rows} ->
-            Nodes = [binary_to_atom(X) || {X} <- Rows, X =/= <<>>],
-            {{ok, Nodes}, State}
-        catch Class:Reason:Stacktrace ->
-                ?LOG_ERROR(#{
-                    what => discovery_failed_select,
-                    class => Class,
-                    reason => Reason,
-                    stacktrace => Stacktrace
-                }),
-                {{error, Reason}, State}
+    try
+        {Num, Nodes, Info} = try_register(ClusterName, Node),
+        mongoose_node_num:set_node_num(Num),
+        {{ok, Nodes}, State#{last_query_info => Info}}
+    catch Class:Reason:Stacktrace ->
+            ?LOG_ERROR(#{what => discovery_failed_select, class => Class,
+                         reason => Reason, stacktrace => Stacktrace}),
+            {{error, Reason}, State}
     end.
 
+try_register(ClusterName, NodeBin) when is_binary(NodeBin), is_binary(ClusterName) ->
+    Node = binary_to_atom(NodeBin),
+    prepare(),
+    {selected, Rows} = select(ClusterName),
+    Pairs = [{binary_to_atom(NodeBin), Num} || {NodeBin, Num} <- Rows],
+    {Nodes, Nums} = lists:unzip(Pairs),
+    AlreadyRegistered = lists:member(Node, Nodes),
+    Timestamp = timestamp(),
+    NodeNum =
+        case AlreadyRegistered of
+            true ->
+                 update_existing(ClusterName, NodeBin, Timestamp),
+                 {value, {_, Num}} = lists:keysearch(Node, 1, Pairs),
+                 Num;
+            false ->
+                 Num = next_free_num(lists:usort(Nums)),
```

Comment:
Why not just have the num as the auto-incremented primary key, and thus leverage the existing functionality already provided by RDBMS? I don't think that avoiding gaps makes a big difference here.

Reply:
It will overflow fast in case we use something that spawns new nodes with new names all the time. 
i.e. for example, autoscaling groups or k8s with random node names.

Autoincrement does not reuse removed values.
The limit for node_num is 255.

The other question: but we don't clean dead nodes from the table to begin with, but we could start doing it though.
I was thinking to make a story to clean nodes with timestamp not updated for some long time.

Also another question: what would happen if there are more than 255 nodes registered? it will probably do something like
```
<<400>>.
<<144>>
```
but it is something to investigate.

Reply:
Thanks for the explanation, we can keep the implementation from this PR.